python/ray/dag/tests/experimental/test_detect_deadlock_dag.py

```diff
@@ -0,0 +1,293 @@
+# coding: utf-8
+import os
+import sys
+
+import pytest
+
+import ray
+import ray.cluster_utils
+from ray.experimental.channel.torch_tensor_type import TorchTensorType
+from ray.experimental.channel.conftest import start_nccl_mock
+from ray.tests.conftest import *  # noqa
+from ray.dag import InputNode, MultiOutputNode
+
+INVALID_GRAPH = "This DAG cannot be compiled because it will "
+"deadlock on NCCL calls. If this is a false positive, please "
+"disable the graph verification by setting the environment "
+"variable RAY_ADAG_ENABLE_VERIFY_GRAPH to 0."
+
+
+@ray.remote(num_cpus=0, num_gpus=1)
+class MockedWorker:
+    def __init__(self):
+        pass
+
+    def start_mock(self):
+        """
+        Patch methods that require CUDA.
+        """
+        start_nccl_mock()
+
+    def no_op(self, value):
+        return value
+
+    def no_op_two(self, value1, value2):
+        return value1, value2
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_gpus": 1}], indirect=True)
+@pytest.mark.parametrize(
+    "tensor_transport", [TorchTensorType.AUTO, TorchTensorType.NCCL]
+)
+def test_invalid_graph_1_actor(ray_start_regular, tensor_transport):
+    """
+    If tensor_transport is TorchTensorType.AUTO, the shared memory channel will be
+    used, and the graph is valid. If tensor_transport is TorchTensorType.NCCL, the
+    NCCL channel will be used, and the graph is invalid.
+
+    [Case: TorchTensorType.NCCL]
+    The first a.no_op writes to the second a.no_op via the NCCL channel. However,
+    the NCCL channel only supports synchronous communication and an actor can
+    only execute one task at a time, so the graph is deadlocked.
```

Comment:
> how the deadlock is detected?

Because the second no_op's in-degree will never be 0, it cannot be included in `zero_in_degree_nodes`, resulting in the non-existence of a topological order for this graph.

Reply:
Ok, sounds good. Is the graph I created above (with the cycle) a correct representation for the scenario?

Reply:
> Is the graph I created above (with the cycle) a correct representation for the scenario?

Yes, it is correct, but you don't add the driver to your graph.